mdata/write_buffer.go

```diff
@@ -0,0 +1,84 @@
+package mdata
+
+import (
+	"gopkg.in/raintank/schema.v1"
+)
+
+/*
+ * The write buffer keeps a window of data during which it is ok to send data out of order.
+ * Once the reorder window has passed it will try to flush the data out.
+ * The write buffer itself is not thread safe because it is only used by AggMetric,
+ * which is thread safe, so there is no locking in the buffer.
+ */
+
+type ReorderBuffer struct {
+	len      uint32                // length of buffer in number of datapoints
+	newest   uint32                // index of newest buffer entry
+	interval uint32                // metric interval
+	buf      []schema.Point        // the actual buffer holding the data
+	flush    func(uint32, float64) // flushCount callback
```

Comment:
FWIW just had a chat on #performance on gophers slack
(https://gophers.slack.com/archives/C0VP8EF3R/p1499947031935022)
and function variables are internally uintptr's so they don't cause GC overhead, however they will take 8B of space. so if we have 10M metrics with each a RB then that's 80MB of pointers all pointing to the same memory location, so yeah the `Process()` idiom seems nicer to me too.